### DISCUSSION TITLE
Documentation to add an automatic action after a file is uploaded

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#20515] [Feature Requests]
### Summary

The goal is to extract data from a CSV file and import them in a collection.
This may require to reformat and filter the data.

### Motivation

To have an automatic flow for the end-user

### Detailed Design

Maybe 
- an interface extension for choosing the file, containing code to transform it and importing the data to the collection
or 
- a flow

### Must Have:
- A
- B

### Should Have:
- C

### Could Have:
- D

### Won't Have:
- E

### Drawbacks

None

### Alternatives

Maybe done with
- an interface extension for choosing the file, containing code to transform it and importing the data to the collection?
or 
- a flow?

### Adoption Strategy

None

**Comment by br41nslug** [comment]
I fail to see why you couldn't already do this using the `files.upload` `action` with a hook extension or Flow 🤔

**Reply by Pachat** [reply]
@br41nslug Thanks.
That's what I thought at first. But in the flows if I have access to the metadata of the file, do I have access to its content?
Because as for using a javascript operation, I read in the doc `"The script will be executed securely in an isolated sandbox. No interaction takes place between the sandbox and the host except for sharing input and output values. This means, for example, no access to the file system and no ability to perform network requests."`
So how can I act on the contents of the file?

**Reply by br41nslug** [reply]
Either using a hook extension or creating a custom operation if you really want to do something with the contents in a Flow.

**Reply by Pachat** [reply]
@br41nslug Thank for your indications. Trying a custom operation, I found out that it is not possible to require `fs`. 
Hence this code doesn't work
```
const fs = require('fs').promises;
export default {
    id: 'cust',
    handler: ({ text }) => {
        console.log(text); // {{$trigger.payload.filename_disk}}
        const data = await fs.readFile('path_to_library'+text, 'utf8');
        console.log(data);
        return 'Data has been read: ' + data;
    },
};
```
How can I read an uploaed file content inside an operation?

**Reply by br41nslug** [reply]
on the api side you should definitely be able to **import** "fs" (best to avoid CommonJS as Directus is writting in ESM). Just the Run Script operation is sandboxed, custom operations put you directly in the nodejs context.

**Reply by Pachat** [reply]
OK. But with this code, there is no flow log at all
```
import { readFileSync } from "fs"
export default {
    id: 'cust',
    handler: ({ text }) => {
        console.log(text); // {{$trigger.payload.filename_disk}}
        const data = readFileSync('path_to_library'+text, 'utf8');
        console.log(data);
        return 'Data has been read: ' + data;
    },
};
```

**Reply by domdus** [reply]
@Pachat some ideas:

- I think the return needs to be of type JSON in order to show up in flows log:
  `return {
      "mydata": data
   }`
   console.log() will only appear in the directus app log (not in flow log)

- For ease of debugging these operations I run the code with local node purely like this:
    ```
    // Add this at the bottom of your script
    // Uncomment for local testing only!

    import test from './api.js'; // import default export
    test.handler({ text: "path_to_your_file" }); // call handler with the param you expect from flow
    ```
    Run it on your local dev env with node command (project root dir):
    `node src/api.js`

   After you're done with debugging and all looks good, comment out the added lines. Then you do all the `npm run build` + move dist files to your directus etc...

--------
Full code:
```
import { readFileSync } from "fs";
export default {
    id: 'cust',
    handler: ({ text }) => {
        console.log(text); // {{$trigger.payload.filename_disk}}
        const data = readFileSync('path_to_library'+text, 'utf8');
        console.log(data);
        return {
          "mydata": data
        };
    },
};

// Uncomment for local testing only!
import test from './api.js'; // import default export
test.handler({ text: "path_to_your_file" }); // call handler with the param you expect from flow
```

**Reply by Pachat** [reply]
@domdus Thank you for your indications and specially for the debugging ones.
I was seeking this for long.
For `fs` isn't it already installed with node?

**Reply by domdus** [reply]
Oh yeah true `fs` should be available. It applies for external node_modules only. (Post updated)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
